File: leetcode/1218/solution.py
"""
https://leetcode.com/problems/longest-arithmetic-subsequence-of-given-difference/
1218. Longest Arithmetic Subsequence of Given Difference

Given an integer array arr and an integer difference, return the length of the longest subsequence in arr which is an arithmetic sequence such that the difference between adjacent elements in the subsequence equals difference.

A subsequence is a sequence that can be derived from arr by deleting some or no elements without changing the order of the remaining elements.
"""
from typing import List, Dict


class Solution:
    def longestSubsequence(self, arr: List[int], difference: int) -> int:
        dp: Dict[int, int] = {}
        longest = 0

        for n in arr:
            prev_value = n - difference
            value = dp.get(prev_value, 0) + 1
            dp[n] = value
            longest = max(longest, value)

        return longest

File: leetcode/1218/solution.rs
impl Solution {
    pub fn longest_subsequence(arr: Vec<i32>, difference: i32) -> i32 {
        use std::collections::HashMap;
        let mut hm = HashMap::new();
        let mut longest = 0;

        for value in arr {
            let prev_value = value - difference;
            let counts = if let Some(v) = hm.get(&prev_value) {
                v + 1
            } else {
                1
            };
            hm.insert(value, counts);
            longest = longest.max(counts);
        }

        longest
    }
}
